### DISCUSSION TITLE
Parsing Data with Multiple Related Tables

### DISCUSSION THREAD

**Discussion by tmathers** [simolus3/drift#2253] [Q&A]
I'm trying write a function to retrieve all records with some many-to-many joins. I want to do it in the least inefficient way possible, so I'm trying to do it with a single query. The documentation only shows examples using `watch()`, but I don't need to watch the data for changes.

What I have currently works, but only a small number of the records are returned. I know it probably has something to do with async/await, but I just can't wrap my head around how to get it to work properly. Now I'm also realizing that I also have to manage duplicates in the joined data, but I can figure that out later.

What's the best way to do with when you have multiple related tables? It seems horribly inefficient to do separate queries for each of them.

```
  Future<List<DrinkData>> getDrinks() async {
    final query = select(drinks).join([
      innerJoin(
          drinksIngredients, drinksIngredients.drinkId.equalsExp(drinks.id)),
      innerJoin(
          ingredients, drinksIngredients.ingredientId.equalsExp(ingredients.id))
      // more joins
    ]);

    final drinkDataMap = <String, DrinkData>{};

    await query.map((row) {
      Drink drink = row.readTable(drinks);

      DrinkData newDrink = new DrinkData(drink);
      drinkDataMap.putIfAbsent(drink.id, () => newDrink);

      IngredientData ingredient = new IngredientData(row.readTable(ingredients));
      drinkDataMap[drink.id].ingredients.add(ingredient);
      // more processing

    }).get();

    return drinkDataMap.values.toList();
  }
```

**Comment by simolus3** [comment]
Actually, neither the query nor the general strategy of mapping rows back looks wrong to me. If you just run `final rows = await query.get()`, do you get the expected amount of rows back or is something missing from there already?

An inner join only works if both result sets have a matching row - if you have joins that should be "optional", you'd need to use a `leftOuterJoin` for that. Do you think any of your joins could be affected by t hat?

**Reply by tmathers** [reply]
You were right, one of the joins was optional, changing to `leftOuterJoin` fixed it. Thanks!

**Reply by tmathers** [reply]
Actually, nevermind. It seemed like that fixed it, because using `leftOuterJoin` returned all of the drink rows, but then the joined table data was missing. I double checked and there aren't any FKs that have null values. And the number of row returned from `query.get()` is still less than it should be.

**Reply by simolus3** [reply]
That's pretty weird. Do you have a way to share the broken code with me so that I can take a deeper look at what's going on?

You could also try to set `logStatements: true` on the `NativeDatabase` constructor you're using to open the database - drift should then print the SQL for all running statements - perhaps there's something weird there? It might also be helpful to pull the database file from the device/emulator and inspect it locally with `sqlite3` to see if all the rows are returned there.

**Reply by tmathers** [reply]
Yeah sorry, turns out that SQLite doesn't work quite the way I thought it did. I think the problem stemmed from the fact that some of the FK's pointed to entries that didn't actually exist. I got it working now!
